src/lib.rs

```diff
@@ -540,3 +543,62 @@ pub trait IntoParams {
         parameter_in_provider: impl Fn() -> Option<openapi::path::ParameterIn>,
     ) -> Vec<openapi::path::Parameter>;
 }
+
+/// This trait is implemented to document a type (like an enum) which can represent multiple
+/// responses, to be used in an operation.
+///
+/// # Examples
+///
+/// ```
+/// use std::collections::BTreeMap;
+/// use utoipa::{
+///     openapi::{Response, ResponseBuilder, ResponsesBuilder},
+///     IntoResponses,
+/// };
+///
+/// enum MyResponse {
+///     Ok,
+///     NotFound,
+/// }
+///
+/// impl IntoResponses for MyResponse {
+///     fn responses() -> BTreeMap<String, Response> {
+///         ResponsesBuilder::new()
+///             .response("200", ResponseBuilder::new().description("Ok"))
+///             .response("404", ResponseBuilder::new().description("Not Found"))
+///             .build()
+///             .into()
+///     }
+/// }
+/// ```
+pub trait IntoResponses {
+    /// Returns an ordered map of response codes to responses.
+    fn responses() -> BTreeMap<String, Response>;
+}
+
+/// This trait is implemented to document a type which represents a single response which can be
+/// referenced or reused as a component in multiple operations.
+///
+/// # Examples
+///
+/// ```
+/// use utoipa::{
+///     openapi::{Response, ResponseBuilder},
+///     ResponseComponent,
+/// };
+///
+/// struct MyResponse;
+///
+/// impl ResponseComponent for MyResponse {
+///     fn response() -> (String, Response) {
+///         (
+///             "MyResponse".to_string(),
+///             ResponseBuilder::new().description("My Response").build(),
+///         )
+///     }
+/// }
+/// ```
+pub trait ResponseComponent {
```

Comment:
Just wondering should this actually just be `Response` instead of a `ResponseComponent`? Component response seems quite long name feels like little bit stuttering? Although if the name is just a `Response` it easily will conflict with other crates which offer types with same name. 

Also wondering should the `Component` derive trait be renamed to `Schema` but it might be too big breaking change for people already using the library. Basically they would have to change it to every type they previously had `Component`

Reply:
I was also tempted by this change, it would be good to make for the sake of clarity and consistency, but perhaps too big of a break? At least with renames it's usually not too difficult to fix with a grep + sed style find replace.

Reply:
> Just wondering should this actually just be Response instead of a ResponseComponent? Component response seems quite long name feels like little bit stuttering? Although if the name is just a Response it easily will conflict with other crates which offer types with same name.

Sure I would be happy to rename it. If it's a conflict people can import it with a rename or with a module scope :slightly_smiling_face:

Reply:
Let's do it in a subsequent PR, seeing as it it is a rather large change!

Reply:
Good

Reply:
> Sure I would be happy to rename it. If it's a conflict people can import it with a rename or with a module scope

The problem here is that it's always a conflict because to implement this trait you must always use `openapi::Response`, the same problem happens with the `Component` trait.  I'm thinking it's still good to keep this as `ResponseComponent`, and rename the `Component` trait to `SchemaComponent` in the future.

Reply:
I'd still prefer them to be one word if possible. Currently the `Component` trait conflicts with the `Component` type. But it is hardly visible to the users. Not sure whether the the `Component` struct should be renamed to `Schema` derivent after renaming `Component` trait. Using `ResponseComponent`or `SchemaComponent`seems little bit of stuttering to me and superfluous to repeat the `Component` word everywhere as well it is unnecessary long to type. :thinking: Is it that it will conflict with the types from the warp or some other?

Reply:
> The problem here is that it's always a conflict because to implement this trait you must always use openapi::Response, the same problem happens with the Component trait.

Typically this won't show up to the users and is not an issue for end users. :thinking:  Since mostly noone will implement manually those traits. Or we could have the traits named as `ResponseComponent` and `SchemaComponent` but the derive name would be `Response` and `Schema`? Was this what you were after? 

> and rename the Component trait to SchemaComponent in the future.

Yeah lets rename that in separete PR so this will not grow overly large.

Also I was considering whether the openapi was defined like this
```rust
#[derive(OpenApi)]
#[openapi(
    schemas(...),
    responses(...),
)]
```
Instead of this. It would be more straight forward but this follows better the OpenAPI spec. 
```rust
#[derive(OpenApi)]
#[openapi(
    components(
        schemas(...),
        responses(...),
    )
)]
```

Also the `handlers` in `#[openapi]` should be renamed to `paths` to follow the spec as well. But that can be done in future PR:s as well.

Reply:
>  Or we could have the traits named as ResponseComponent and SchemaComponent but the derive name would be Response and Schema?

Sure I can support that :slightly_smiling_face: Having some distinguishing feature in the name other than one belonging to the `openapi` module and the other not, because that's not enough for me personally to guess at a glance that they represent different concepts. 

> Instead of this. It would be more straight forward but this follows better the OpenAPI spec.

That looks good, do you think it might be good to represent this structure in the underlying data types using something like a `ComponentsBuilder`?

> Also the handlers in #[openapi] should be renamed to paths to follow the spec as well. But that can be done in future PR:s as well.

Great idea!

Reply:
> Or we could have the traits named as ResponseComponent and SchemaComponent but the derive name would be Response and Schema?

Actually on second thoughts, shouldn't the derive name match the trait? That seems to be standard practice in the Rust community.

Reply:
In a sense, the trait `Component` isn't actually a component, it's a component generator, it generates openapi components. Perhaps `ComponentGenerator` or something with a similar meaning would be a better name for it?

Reply:
> Actually on second thoughts, shouldn't the derive name match the trait? That seems to be standard practice in the Rust community.

Yeah I was wondering the same actually so better name them same :) 

> In a sense, the traits Schema (as proposed) isn't actually a , it's a schema generator, it generates openapi schema on demand. Perhaps SchemaGenerator or something with a similar meaning would be a better name for it? Same deal for Response trait.

Maybe `ToSchema` and `ToResponse` this would then follow same convention as syn crate's `ToTokens`?

Reply:
> That looks good, do you think it might be good to represent this structure in the underlying data types using something like a ComponentsBuilder?

If something then the `Comonent` references should be probably renamed to `Schema`s. And similarly the function `components_from_iter` and `component` should be renamed to `schemas_from_iter` and to `schema`. Then it would follow the spec more closely.

Reply:
> Perhaps `SchemaGenerator` or something with a similar meaning

Yeah the `SchemaGenerator` indicate that it creates a generator object. Perhaps `GenerateSchema` but again it would be unnecessary long or `GenSchema`? Like serde's `Serialize` or `Deserialize` what would be if possible one word or as short as possible descriptive action that it would perform. Rust prefers naming traits as one verb but also `or` and `er` endings are somewhat common too. I guess `ToSchema` and `ToResponse` would be not too far fetched and sound quite  reasonable?

Reply:
> components_from_iter and component should be renamed to schemas_from_iter and to schema. Then it would follow the spec more closely.

I've made that change in the PR for `ComponentsBuilder`, because it stores the components in a variable called `schemas` already, so I made it match.

Reply:
> Yeah the `SchemaGenerator` indicate that it creates a generator object. Perhaps `GenerateSchema` but again it would be unnecessary long or `GenSchema`? Like serde's `Serialize` or `Deserialize` what would be if possible one word or as short as possible descriptive action that it would perform. Rust prefers naming traits as one verb but also `or` and `er` endings are somewhat common too. I guess `ToSchema` and `ToResponse` would be not too far fetched and sound quite reasonable?

I think using the full word is usually better than shortcuts because it makes the code more accessible to those who don't speak English natively (and are using translation to understand), or those who are using screen readers. We should optimize for reading. But `ToSchema` and `ToResponse` seem like a reasonable option to me, let's go with that. I've renamed the `ResponseComponent` to `ToResponse`. I can do the `Component` to `ToSchema` rename in the next PR. 

Is this PR ready now do you think?

Reply:
Yeah you are right.. Let's get the ball rolling and get this thing to the upstream. Thanks for your work :heart: